v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,247 @@
+    enum { _tlgTagConst = (0) };
+    enum { _tlgLevelConst = 5 & 0 | (5) };
+
+    TraceLoggingHProvider const _tlgProv = (g_my_component_provider);
+    if (trace_level < _tlgProv->LevelPlus1 && _tlgKeywordOn(_tlgProv, 0))
+    {
+        // have a stack allocated buffer where we construct the event metadata and
+        uint8_t _tlgEvent[MAX_SELF_DESCRIBED_EVENT_WITH_METADATA];
+
+        // compute event metadata size
+        size_t event_name_length = strlen(event_name);
+        uint16_t metadata_size = (uint16_t)(event_name_length + 1);
+        metadata_size += sizeof("content") + 1 + sizeof("file") + 1 + sizeof("func") + 1 + sizeof("line") + 1;
+        for (uint32_t i = 0; i < property_value_count; i++)
+        {
+            size_t name_length = strlen(context_property_value_pairs[i].name);
+            metadata_size += (uint16_t)name_length + 1 + 1;
+        }
+
+        // alias to the event bytes
+        SELF_DESCRIBED_EVENT* self_described_event = (SELF_DESCRIBED_EVENT*)_tlgEvent;
+        self_described_event->_tlgBlobTyp = _TlgBlobEvent4;
+        self_described_event->_tlgChannel = 11;
+        self_described_event->_tlgLevel = _tlgLevelConst;
+        self_described_event->_tlgOpcode = 0;
+        self_described_event->_tlgKeyword = 0;
+        self_described_event->_tlgEvtMetaSize = metadata_size + 4;
+        self_described_event->_tlgEvtTag = 128 | ((0xfe00000 & (UINT32)_tlgTagConst) >> 21) | ((0x01fc000 & (UINT32)_tlgTagConst) >> 6) | ((0x0003fff & (UINT32)_tlgTagConst) << 16) | (~(0x0FFFFFFF | ~(UINT32)_tlgTagConst));
+
+        // at this point we filled all the event information, now we need to provide the metadata bytes
+        // first one is the event name, followed by metadata for all the fields
+        // copy event name 
+        uint8_t* pos = &self_described_event->metadata[0];
+        (void)memcpy(pos, event_name, event_name_length + 1); pos += event_name_length + 1;
+
+        // copy the field metadata (name and in type)
+        (void)memcpy(pos, "content", sizeof("content")); pos += sizeof("content");
+        *pos = TlgInANSISTRING;  pos++;
+        (void)memcpy(pos, "file", sizeof("file")); pos += sizeof("file");
+        *pos = TlgInANSISTRING;  pos++;
+        (void)memcpy(pos, "func", sizeof("func")); pos += sizeof("func");
+        *pos = TlgInANSISTRING;  pos++;
+        (void)memcpy(pos, "line", sizeof("line")); pos += sizeof("line");
+        *pos = TlgInINT32;  pos++;
+        for (uint32_t i = 0; i < property_value_count; i++)
+        {
+            size_t name_length = strlen(context_property_value_pairs[i].name);
+            (void)memcpy(pos, context_property_value_pairs[i].name, name_length + 1); pos += name_length + 1;
+            switch (context_property_value_pairs[i].type->get_type())
+            {
+            default:
+                *pos = TlgInANSISTRING;
+                break;
+            case LOG_CONTEXT_PROPERTY_TYPE_int32_t:
```

Comment:
>LOG_CONTEXT_PROPERTY_TYPE_int32_t [](http://example.com/codeflow?start=17&length=33)

likely needs more types here. I find all this code fascinating.

I cannot even imagine the amount of reverse engineering it took.

Reply:
Absolutely, plan is to support almost all the ETW types (at least bool and GUID would be nice additions).
And it was quite fun. The kind of fun where you hit a brick wall that you can't do something and then you're like: REAAAALLY? I can't do what? Challenge accepted, after all it is all bytes! So I cannot deny that I really enjoyed it...

Reply:
Supporting all these will come as a separate PR (after the .md is merged in.